`crucible-llvm`: Support the `llvm.{s,u}cmp.*` intrinsics

Clang 20 or later is more likely to optimize three-way comparisons into uses the [`llvm.scmp.*`](https://releases.llvm.org/20.1.0/docs/LangRef.html#llvm-scmp-intrinsic) and [`llvm.ucmp.*`](https://releases.llvm.org/20.1.0/docs/LangRef.html#llvm-ucmp-intrinsic) intrinsics. Here is an example of a C program which emits such intrinsics when compiled with optimizations:

```c
// test.c
#include <stdint.h>

__attribute__((noinline)) uint32_t ucmp(uint32_t a, uint32_t b) {
    return a == b ? 0 : (a < b ? -1 : 1);
}

__attribute__((noinline)) int32_t scmp(int32_t a, int32_t b) {
    return a == b ? 0 : (a < b ? -1 : 1);
}

int main(void) {
    return ucmp(0, 0) || scmp(0, 0);
}
```

`crucible-llvm` does not currently support these intrinsics, as can be seen if you try to simulate this program using `crux-llvm`:

<details>

```
$ CLANG=clang-20 LLVM_LINK=llvm-link-20 cabal -v0 run exe:crux-llvm -- test.c
[Crux] Using pointer width: 64 for file crux-build/crux~test.bc
[Crux] Simulating function main
[Crux] Attempting to prove verification conditions.
[Crux] *** debug executable: results/test/debug-5
[Crux] *** break on line: 5
[Crux] Found counterexample for verification goal
[Crux]   test.c:5:12: error: in ucmp
[Crux]   Failed to load function handle
[Crux]   Details:
[Crux]     No implementation or override found for pointer
[Crux]     The given pointer could not be resolved to a callable function
[Crux]     No implementation or override found for pointer
[Crux]     Attempting to load callable function with type: i32(i32, i32)
[Crux]       Via pointer: Global symbol "llvm.ucmp.i32.i32" (1, 0x0:[64])
[Crux]     In memory state:
[Crux]       Stack frame ucmp
[Crux]         No writes or allocations
[Crux]       Stack frame main
[Crux]         No writes or allocations
[Crux]       Base memory
[Crux]         Allocations:
[Crux]           GlobalAlloc 5 0x0:[64] Immutable 1-byte-aligned [defined function ] main
[Crux]           GlobalAlloc 4 0x0:[64] Immutable 1-byte-aligned [defined function ] scmp
[Crux]           GlobalAlloc 3 0x0:[64] Immutable 1-byte-aligned [defined function ] ucmp
[Crux]           GlobalAlloc 2 0x0:[64] Immutable 1-byte-aligned [external function] llvm.scmp.i32.i32
[Crux]           GlobalAlloc 1 0x0:[64] Immutable 1-byte-aligned [external function] llvm.ucmp.i32.i32
[Crux]     in context:
[Crux]       ucmp
[Crux]       main
[Crux] Goal status:
[Crux]   Total: 1
[Crux]   Proved: 0
[Crux]   Disproved: 1
[Crux]   Incomplete: 0
[Crux]   Unknown: 0
[Crux] Overall status: Invalid.
```
</details>

Each intrinsic returns `-1` if the first argument is less than the second argument, `0` if the arguments are equal, and `1` if the first argument is greater than second argument.

--- a/crucible-llvm-cli/test-data/llvm.scmp.i2.i32.cbl
+++ b/crucible-llvm-cli/test-data/llvm.scmp.i2.i32.cbl
@@ -1,0 +1,13 @@
+(declare @llvm.scmp.i2.i32 ((x (Bitvector 32)) (y (Bitvector 32))) (Bitvector 2))
+
+(defun @main () Unit
+  (start start:
+    (let x (bv 32 -1))
+    (let y (bv 32 1))
+    (let r0 (funcall @llvm.scmp.i2.i32 x y))
+    (let r1 (funcall @llvm.scmp.i2.i32 x x))
+    (let r2 (funcall @llvm.scmp.i2.i32 y x))
+    (assert! (equal? r0 (bv 2 -1)) "r0 == -1")
+    (assert! (equal? r1 (bv 2 0)) "r1 == 0")
+    (assert! (equal? r2 (bv 2 1)) "r2 == 1")
+    (return ())))

--- a/crucible-llvm-cli/test-data/llvm.scmp.i2.i32.out.good
+++ b/crucible-llvm-cli/test-data/llvm.scmp.i2.i32.out.good
@@ -1,0 +1,4 @@
+==== Begin Simulation ====
+
+==== Finish Simulation ====
+==== No proof obligations ====

--- a/crucible-llvm-cli/test-data/llvm.ucmp.i2.i32.cbl
+++ b/crucible-llvm-cli/test-data/llvm.ucmp.i2.i32.cbl
@@ -1,0 +1,13 @@
+(declare @llvm.ucmp.i2.i32 ((x (Bitvector 32)) (y (Bitvector 32))) (Bitvector 2))
+
+(defun @main () Unit
+  (start start:
+    (let x (bv 32 1))
+    (let y (bv 32 2))
+    (let r0 (funcall @llvm.ucmp.i2.i32 x y))
+    (let r1 (funcall @llvm.ucmp.i2.i32 x x))
+    (let r2 (funcall @llvm.ucmp.i2.i32 y x))
+    (assert! (equal? r0 (bv 2 -1)) "r0 == -1")
+    (assert! (equal? r1 (bv 2 0)) "r1 == 0")
+    (assert! (equal? r2 (bv 2 1)) "r2 == 1")
+    (return ())))

--- a/crucible-llvm-cli/test-data/llvm.ucmp.i2.i32.out.good
+++ b/crucible-llvm-cli/test-data/llvm.ucmp.i2.i32.out.good
@@ -1,0 +1,4 @@
+==== Begin Simulation ====
+
+==== Finish Simulation ====
+==== No proof obligations ====

--- a/crucible-llvm/CHANGELOG.md
+++ b/crucible-llvm/CHANGELOG.md
@@ -17,6 +17,7 @@
     * Replace `llvmOverride_name` with `llvmOvSymbol`
     * Replace `llvmOverride_args` with `llvmOvArgs`
     * Replace `llvmOverride_ret` with `llvmOvRet`
+* Support the `llvm.scmp.*` and `llvm.ucmp.*` three-way comparison intrinsics.
 
 # 0.9 -- 2026-01-29
 

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics.hs
@@ -187,6 +187,7 @@ declare_overrides =
   , map (\(SomeLLVMOverride ov) -> basic_llvm_override ov) LLVM.basic_llvm_overrides
   , map (\(pfx, LLVM.Poly1LLVMOverride ov) -> polymorphic1_llvm_override pfx ov) LLVM.poly1_llvm_overrides
   , map (\(pfx, LLVM.Poly1VecLLVMOverride ov) -> polymorphic1_vec_llvm_override pfx ov) LLVM.poly1_vec_llvm_overrides
+  , map (\(pfx, LLVM.PolyCmpLLVMOverride ov) -> polymorphic_cmp_llvm_override pfx ov) LLVM.poly_cmp_llvm_overrides
 
   -- C++ standard library functions
   , [ Libcxx.register_cpp_override Libcxx.endlOverride ]

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Common.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Common.hs
@@ -34,6 +34,7 @@ module Lang.Crucible.LLVM.Intrinsics.Common
   , basic_llvm_override
   , polymorphic1_llvm_override
   , polymorphic1_vec_llvm_override
+  , polymorphic_cmp_llvm_override
 
   , llvmOverrideToTypedOverride
   , register_llvm_override
@@ -212,6 +213,29 @@ polymorphic1_vec_llvm_override :: forall p sym ext arch wptr.
 polymorphic1_vec_llvm_override prefix fn =
   OverrideTemplate (Match.PrefixMatch prefix) (register_1arg_vec_polymorphic_override prefix fn)
 
+-- | Create an 'OverrideTemplate' for an intrinsic in the @llvm.{s,u}cmp.*@
+-- family. These can be instantiated at multiple types, including:
+--
+-- * @i2 \@llvm.scmp.i2.i32(i32, i32)@
+
+-- * @i8 \@llvm.scmp.i8.i8(i8, i8)@
+--
+-- * etc.
+--
+-- Note that the argument and result types are allowed to be different, so the
+-- @fn@ argument is parameterized by two separate 'NatRepr's.
+polymorphic_cmp_llvm_override :: forall p sym ext arch wptr.
+  (IsSymInterface sym, HasLLVMAnn sym, HasPtrWidth wptr) =>
+  String ->
+  (forall argSz resSz.
+    (1 <= argSz, 2 <= resSz) =>
+    NatRepr argSz ->
+    NatRepr resSz ->
+    SomeLLVMOverride p sym ext) ->
+  OverrideTemplate p sym ext arch
+polymorphic_cmp_llvm_override prefix fn =
+  OverrideTemplate (Match.PrefixMatch prefix) (register_cmp_polymorphic_override prefix fn)
+
 register_1arg_polymorphic_override :: forall p sym ext arch wptr.
   (IsSymInterface sym, HasLLVMAnn sym, HasPtrWidth wptr) =>
   String ->
@@ -255,6 +279,36 @@ register_1arg_vec_polymorphic_override prefix overrideFn =
         , Some intSzRepr <- mkNatRepr intSzNat
         , Just LeqProof <- isPosNat intSzRepr
         -> Just (overrideFn vecSzRepr intSzRepr)
+      _ -> Nothing
+
+-- | Register an override for an intrinsic in the @llvm.{s,u}cmp.*@ family.
+-- (See the Haddocks for 'polymorphic_cmp_llvm_override' for details on what
+-- this means.) This function is responsible for parsing the suffixes in the
+-- intrinsic's name, which encodes the sizes of the argument and result types.
+-- As some examples:
+--
+-- * @.i2.i32@ (argument size @32@, result size @2@)
+--
+-- * @.i8.i64@ (argument size @64@, result size @8@)
+register_cmp_polymorphic_override :: forall p sym ext arch wptr.
+  (IsSymInterface sym, HasLLVMAnn sym, HasPtrWidth wptr) =>
+  String ->
+  (forall argSz resSz.
+    (1 <= argSz, 2 <= resSz) =>
+    NatRepr argSz ->
+    NatRepr resSz ->
+    SomeLLVMOverride p sym ext) ->
+  MakeOverride p sym ext arch
+register_cmp_polymorphic_override prefix overrideFn =
+  MakeOverride $ \(Decl.SomeDeclare (Decl.Declare{ Decl.decName = L.Symbol nm })) _ _ ->
+    case List.stripPrefix prefix nm of
+      Just ('.':'i': (readDec -> (resSz,rest):_))
+        | '.':'i': (readDec -> (argSz,[]):_) <- rest
+        , Some argW <- mkNatRepr argSz
+        , Some resW <- mkNatRepr resSz
+        , Just LeqProof <- testLeq (knownNat @1) argW
+        , Just LeqProof <- testLeq (knownNat @2) resW
+        -> Just (overrideFn argW resW)
       _ -> Nothing
 
 basic_llvm_override :: forall p args ret sym ext arch wptr.

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
@@ -344,6 +344,35 @@ poly1_vec_llvm_overrides =
     )
   ]
 
+-- | An LLVM override for the @llvm.{s,u}cmp.*@ family of intrinsics. These
+-- overrides have a unique shape in that:
+--
+-- 1. Both their argument and result types are polymorphic integer types (which
+--    are allowed to be different), and
+--
+-- 2. The result integer type must have a size of at least 2 bits.
+newtype PolyCmpLLVMOverride p sym ext
+  = PolyCmpLLVMOverride
+      (forall argSz resSz
+         . (1 <= argSz, 2 <= resSz)
+        => NatRepr argSz
+        -> NatRepr resSz
+        -> SomeLLVMOverride p sym ext)
+
+poly_cmp_llvm_overrides ::
+  IsSymInterface sym =>
+  [(String, PolyCmpLLVMOverride p sym ext)]
+poly_cmp_llvm_overrides =
+  [ ("llvm.scmp"
+    , PolyCmpLLVMOverride $ \argSz resSz ->
+        SomeLLVMOverride (llvmScmp argSz resSz)
+    )
+  , ("llvm.ucmp"
+    , PolyCmpLLVMOverride $ \argSz resSz ->
+        SomeLLVMOverride (llvmUcmp argSz resSz)
+    )
+  ]
+
 ------------------------------------------------------------------------
 -- ** Declarations
 
@@ -1739,6 +1768,54 @@ llvmUminVector ::
      (VectorType (BVType intSz))
 llvmUminVector = llvmVectorMap "umin" callVectorMapUmin
 
+-- | Build an 'LLVMOverride' for an @llvm.{s,u}cmp.*@ intrinsic.
+llvmCmp ::
+  forall argSz resSz p sym ext.
+  (1 <= argSz, 2 <= resSz) =>
+  -- | The name of the operation (@scmp@ or @ucmp@).
+  String ->
+  -- | The semantics of the override.
+  (forall r args ret.
+    IsSymInterface sym =>
+    RegEntry sym (BVType argSz) ->
+    RegEntry sym (BVType argSz) ->
+    OverrideSim p sym ext r args ret (SymBV sym resSz)) ->
+  -- | The size of the argument type.
+  NatRepr argSz ->
+  -- | The size of the result type.
+  NatRepr resSz ->
+  LLVMOverride p sym ext
+     (EmptyCtx ::> BVType argSz ::> BVType argSz)
+     (BVType resSz)
+llvmCmp cmpOpName cmpOp argSz resSz
+  | (LeqProof :: LeqProof 1 resSz) <-
+      leqTrans (LeqProof @1 @2) (LeqProof @2 @resSz)
+  = let nm = L.Symbol ("llvm." ++ cmpOpName ++
+                       ".i" ++ show (natValue resSz) ++
+                       ".i" ++ show (natValue argSz)) in
+      [llvmOvr| #resSz $nm( #argSz, #argSz ) |]
+      (\_memOps args -> Ctx.uncurryAssignment cmpOp args)
+
+llvmScmp ::
+  forall argSz resSz p sym ext.
+  (1 <= argSz, 2 <= resSz) =>
+  NatRepr argSz ->
+  NatRepr resSz ->
+  LLVMOverride p sym ext
+     (EmptyCtx ::> BVType argSz ::> BVType argSz)
+     (BVType resSz)
+llvmScmp argSz resSz = llvmCmp "scmp" (callScmp resSz) argSz resSz
+
+llvmUcmp ::
+  forall argSz resSz p sym ext.
+  (1 <= argSz, 2 <= resSz) =>
+  NatRepr argSz ->
+  NatRepr resSz ->
+  LLVMOverride p sym ext
+     (EmptyCtx ::> BVType argSz ::> BVType argSz)
+     (BVType resSz)
+llvmUcmp argSz resSz = llvmCmp "ucmp" (callUcmp resSz) argSz resSz
+
 ------------------------------------------------------------------------
 -- ** Implementations
 
@@ -2459,3 +2536,46 @@ callVectorMapUmin ::
 callVectorMapUmin vec1 vec2 = do
   sym <- getSymInterface
   callVectorMap (bvUmin sym) vec1 vec2
+
+-- | The semantics of an @llvm.{s,u}cmp.*@ intrinsic.
+callCmp ::
+  forall argSz resSz p sym ext r args ret.
+  (IsSymInterface sym, 1 <= argSz, 2 <= resSz) =>
+  -- | The semantics of a bitvector less-than operation (either signed or
+  -- unsigned).
+  (sym -> SymBV sym argSz -> SymBV sym argSz -> IO (Pred sym)) ->
+  -- | The size of the result type.
+  NatRepr resSz ->
+  RegEntry sym (BVType argSz) ->
+  RegEntry sym (BVType argSz) ->
+  OverrideSim p sym ext r args ret (SymBV sym resSz)
+callCmp bvLtOp resSz (regValue -> x) (regValue -> y)
+  | (LeqProof :: LeqProof 1 resSz) <-
+      leqTrans (LeqProof @1 @2) (LeqProof @2 @resSz)
+  = do sym <- getSymInterface
+       liftIO $
+         do xLtY <- bvLtOp sym x y
+            xEqY <- bvEq sym x y
+            zero <- bvZero sym resSz
+            one <- bvOne sym resSz
+            negOne <- bvNeg sym one
+            -- If x < y, return -1. If x == y, return 0. If x > y, return 1.
+            bvIte sym xLtY negOne =<< bvIte sym xEqY zero one
+
+callScmp ::
+  forall argSz resSz p sym ext r args ret.
+  (IsSymInterface sym, 1 <= argSz, 2 <= resSz) =>
+  NatRepr resSz ->
+  RegEntry sym (BVType argSz) ->
+  RegEntry sym (BVType argSz) ->
+  OverrideSim p sym ext r args ret (SymBV sym resSz)
+callScmp = callCmp bvSlt
+
+callUcmp ::
+  forall argSz resSz p sym ext r args ret.
+  (IsSymInterface sym, 1 <= argSz, 2 <= resSz) =>
+  NatRepr resSz ->
+  RegEntry sym (BVType argSz) ->
+  RegEntry sym (BVType argSz) ->
+  OverrideSim p sym ext r args ret (SymBV sym resSz)
+callUcmp = callCmp bvUlt

